src/io/flutter/run/daemon/DevToolsService.java

```diff
@@ -0,0 +1,134 @@
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
+package io.flutter.run.daemon;
+
+import com.google.common.collect.ImmutableList;
+import com.intellij.execution.ExecutionException;
+import com.intellij.execution.configurations.GeneralCommandLine;
+import com.intellij.execution.process.ProcessHandler;
+import com.intellij.openapi.components.ServiceManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.project.ProjectManager;
+import com.intellij.openapi.project.ProjectManagerListener;
+import com.intellij.openapi.util.io.FileUtil;
+import com.intellij.openapi.vfs.CharsetToolkit;
+import io.flutter.FlutterUtils;
+import io.flutter.android.IntelliJAndroidSdk;
+import io.flutter.bazel.Workspace;
+import io.flutter.bazel.WorkspaceCache;
+import io.flutter.sdk.FlutterSdk;
+import io.flutter.sdk.FlutterSdkUtil;
+import io.flutter.utils.MostlySilentOsProcessHandler;
+import org.jetbrains.annotations.NotNull;
+
+import static com.jetbrains.lang.dart.ide.runner.server.vmService.VmServiceWrapper.LOG;
+
+public class DevToolsService {
+  private class DevToolsServiceListener implements DaemonEvent.Listener {
+  }
+
+  @NotNull private final Project project;
+  private DaemonApi daemonApi;
+  private ProcessHandler process;
+  private DevToolsInstance devToolsInstance;
+
+  @NotNull
+  public static DevToolsService getInstance(@NotNull final Project project) {
+    return ServiceManager.getService(project, DevToolsService.class);
+  }
+
+  private DevToolsService(@NotNull final Project project) {
+    this.project = project;
+    try {
+      final GeneralCommandLine command = chooseCommand(project);
+      assert command != null;
+      this.process = new MostlySilentOsProcessHandler(command);
+      daemonApi = new DaemonApi(process);
+      daemonApi.listen(process, new DevToolsServiceListener());
+      daemonApi.devToolsServe().thenAccept((DaemonApi.DevToolsAddress address) -> {
+        if (!project.isOpen()) {
+          // We should skip starting DevTools (and doing any UI work) if the project has been closed.
+          return;
+        }
+        if (address == null) {
+          LOG.error("DevTools address was null");
+        }
+        else {
+          devToolsInstance = new DevToolsInstance(address.host, address.port);
+        }
+      });
+    }
+    catch (ExecutionException e) {
+      e.printStackTrace();
+    }
+
+    ProjectManager.getInstance().addProjectManagerListener(project, new ProjectManagerListener() {
+      @Override
+      public void projectClosing(@NotNull Project project) {
+        daemonApi.daemonShutdown();
+        if (!process.isProcessTerminated()) {
+          process.destroyProcess();
+        }
+      }
+    });
+  }
+
+  private static GeneralCommandLine chooseCommand(@NotNull final Project project) {
+    final String androidHome = IntelliJAndroidSdk.chooseAndroidHome(project, false);
```

Comment:
why are we passing android home in. That seems like it has nothing to do with serving devtools. We should either share this code with the regular code to launch the daemon or remove code that is unrelated to the launching a daemon just to serve devtools.

Reply:
Ah I didn't understand why this was relevant for DeviceService (where I copied this code from), but I think I can just delete this.